File: articles/self-concept-self-esteem-and-self-knowledge/r/analyze_self_concept_self_esteem_self_knowledge.R
root <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."))
data_path <- file.path(root, "data", "synthetic_self_concept_self_esteem_self_knowledge.csv")
outputs <- file.path(root, "outputs")
dir.create(outputs, showWarnings = FALSE, recursive = TRUE)

data <- read.csv(data_path, stringsAsFactors = FALSE)

agreement_summary <- data.frame(
  domain = c("warmth", "conscientiousness", "emotional_stability", "openness"),
  self_other_agreement = c(
    cor(data$self_warmth, data$other_warmth, use = "pairwise.complete.obs"),
    cor(data$self_conscientiousness, data$other_conscientiousness, use = "pairwise.complete.obs"),
    cor(data$self_emotional_stability, data$other_emotional_stability, use = "pairwise.complete.obs"),
    cor(data$self_openness, data$other_openness, use = "pairwise.complete.obs")
  )
)

data$esteem_level <- ifelse(
  data$self_esteem > median(data$self_esteem),
  "higher_self_esteem",
  "lower_self_esteem"
)
data$accuracy_level <- ifelse(
  data$self_knowledge_accuracy > median(data$self_knowledge_accuracy),
  "higher_self_knowledge",
  "lower_self_knowledge"
)
data$discrepancy_level <- ifelse(
  data$total_self_discrepancy > median(data$total_self_discrepancy),
  "higher_discrepancy",
  "lower_discrepancy"
)
data$self_system_profile <- paste(
  data$esteem_level,
  data$accuracy_level,
  data$discrepancy_level,
  sep = "_"
)

data$high_esteem_low_accuracy <- data$self_esteem > median(data$self_esteem) &
  data$self_knowledge_accuracy < median(data$self_knowledge_accuracy)
data$low_esteem_high_devaluation <- data$self_esteem < median(data$self_esteem) &
  data$external_devaluation > median(data$external_devaluation)

context_summary <- aggregate(
  data[, c("self_concept_positivity", "self_esteem", "self_knowledge_accuracy", "total_self_discrepancy", "social_recognition", "external_devaluation", "well_being")],
  by = list(self_system_context = data$self_system_context),
  FUN = mean
)

profile_summary <- aggregate(
  data[, c("self_esteem", "self_knowledge_accuracy", "total_self_discrepancy", "social_recognition", "external_devaluation", "well_being")],
  by = list(self_system_profile = data$self_system_profile),
  FUN = mean
)

corr_cols <- c(
  "self_warmth", "self_conscientiousness", "self_emotional_stability", "self_openness",
  "other_warmth", "other_conscientiousness", "other_emotional_stability", "other_openness",
  "actual_self", "ideal_self", "ought_self", "self_esteem", "social_recognition",
  "external_devaluation", "well_being", "self_other_gap_mean", "self_knowledge_accuracy",
  "actual_ideal_discrepancy", "actual_ought_discrepancy", "total_self_discrepancy",
  "self_concept_positivity"
)

model_self_esteem <- lm(
  self_esteem ~ self_concept_positivity + total_self_discrepancy +
    self_knowledge_accuracy + social_recognition + external_devaluation,
  data = data
)

model_wellbeing <- lm(
  well_being ~ self_esteem + total_self_discrepancy +
    self_knowledge_accuracy + social_recognition + external_devaluation,
  data = data
)

model_accuracy <- lm(
  self_knowledge_accuracy ~ social_recognition + external_devaluation +
    self_concept_positivity + self_esteem,
  data = data
)

write.csv(context_summary, file.path(outputs, "r_self_system_context_summary.csv"), row.names = FALSE)
write.csv(agreement_summary, file.path(outputs, "r_self_other_agreement_summary.csv"), row.names = FALSE)
write.csv(profile_summary, file.path(outputs, "r_self_system_profile_summary.csv"), row.names = FALSE)
write.csv(cor(data[, corr_cols], use = "pairwise.complete.obs"), file.path(outputs, "r_self_system_correlations.csv"))
capture.output(summary(model_self_esteem), file = file.path(outputs, "r_self_esteem_model_summary.txt"))
capture.output(summary(model_wellbeing), file = file.path(outputs, "r_wellbeing_model_summary.txt"))
capture.output(summary(model_accuracy), file = file.path(outputs, "r_self_knowledge_accuracy_model_summary.txt"))
write.csv(data, file.path(outputs, "r_scored_self_concept_self_esteem_self_knowledge.csv"), row.names = FALSE)
cat("Wrote R outputs to:", outputs, "\n")

File: articles/self-concept-self-esteem-and-self-knowledge/python/analyze_self_concept_self_esteem_self_knowledge.py
#!/usr/bin/env python3
from pathlib import Path
import json
import numpy as np
import pandas as pd

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data" / "synthetic_self_concept_self_esteem_self_knowledge.csv"
OUT = ROOT / "outputs"
OUT.mkdir(exist_ok=True)

SELF_COLS = ["self_warmth", "self_conscientiousness", "self_emotional_stability", "self_openness"]
OTHER_COLS = ["other_warmth", "other_conscientiousness", "other_emotional_stability", "other_openness"]
BASE = SELF_COLS + OTHER_COLS + [
    "actual_self",
    "ideal_self",
    "ought_self",
    "self_esteem",
    "social_recognition",
    "external_devaluation",
    "well_being",
]
DERIVED = [
    "warmth_gap",
    "conscientiousness_gap",
    "emotional_stability_gap",
    "openness_gap",
    "self_other_gap_mean",
    "self_knowledge_accuracy",
    "actual_ideal_discrepancy",
    "actual_ought_discrepancy",
    "total_self_discrepancy",
    "self_concept_positivity",
]

def ols(df, y, xs):
    X = np.column_stack([np.ones(len(df)), df[xs].to_numpy(float)])
    beta, *_ = np.linalg.lstsq(X, df[y].to_numpy(float), rcond=None)
    return dict(zip(["intercept", *xs], map(float, beta)))

df = pd.read_csv(DATA)
required = {"person_id", "self_system_context", *BASE, *DERIVED}
missing = required - set(df.columns)
if missing:
    raise ValueError(f"Missing columns: {sorted(missing)}")

agreement_summary = pd.DataFrame({
    "domain": ["warmth", "conscientiousness", "emotional_stability", "openness"],
    "self_other_agreement": [
        df["self_warmth"].corr(df["other_warmth"]),
        df["self_conscientiousness"].corr(df["other_conscientiousness"]),
        df["self_emotional_stability"].corr(df["other_emotional_stability"]),
        df["self_openness"].corr(df["other_openness"]),
    ],
})

df["esteem_level"] = np.where(df["self_esteem"] > df["self_esteem"].median(), "higher_self_esteem", "lower_self_esteem")
df["accuracy_level"] = np.where(df["self_knowledge_accuracy"] > df["self_knowledge_accuracy"].median(), "higher_self_knowledge", "lower_self_knowledge")
df["discrepancy_level"] = np.where(df["total_self_discrepancy"] > df["total_self_discrepancy"].median(), "higher_discrepancy", "lower_discrepancy")
df["self_system_profile"] = df["esteem_level"] + "_" + df["accuracy_level"] + "_" + df["discrepancy_level"]

df["high_esteem_low_accuracy"] = (
    (df["self_esteem"] > df["self_esteem"].median())
    & (df["self_knowledge_accuracy"] < df["self_knowledge_accuracy"].median())
)
df["low_esteem_high_devaluation"] = (
    (df["self_esteem"] < df["self_esteem"].median())
    & (df["external_devaluation"] > df["external_devaluation"].median())
)

context_summary = (
    df.groupby("self_system_context")
    .agg(
        n=("person_id", "count"),
        self_concept_positivity_mean=("self_concept_positivity", "mean"),
        self_esteem_mean=("self_esteem", "mean"),
        self_knowledge_accuracy_mean=("self_knowledge_accuracy", "mean"),
        total_self_discrepancy_mean=("total_self_discrepancy", "mean"),
        social_recognition_mean=("social_recognition", "mean"),
        external_devaluation_mean=("external_devaluation", "mean"),
        well_being_mean=("well_being", "mean"),
    )
    .reset_index()
)

profile_summary = (
    df.groupby("self_system_profile")
    .agg(
        n=("person_id", "count"),
        self_esteem_mean=("self_esteem", "mean"),
        self_knowledge_accuracy_mean=("self_knowledge_accuracy", "mean"),
        total_self_discrepancy_mean=("total_self_discrepancy", "mean"),
        social_recognition_mean=("social_recognition", "mean"),
        external_devaluation_mean=("external_devaluation", "mean"),
        well_being_mean=("well_being", "mean"),
    )
    .reset_index()
)

pattern_summary = pd.DataFrame({
    "pattern": ["high_esteem_low_accuracy", "low_esteem_high_devaluation"],
    "n": [int(df["high_esteem_low_accuracy"].sum()), int(df["low_esteem_high_devaluation"].sum())],
    "proportion": [float(df["high_esteem_low_accuracy"].mean()), float(df["low_esteem_high_devaluation"].mean())],
})

models = {
    "self_esteem": ols(df, "self_esteem", ["self_concept_positivity", "total_self_discrepancy", "self_knowledge_accuracy", "social_recognition", "external_devaluation"]),
    "well_being": ols(df, "well_being", ["self_esteem", "total_self_discrepancy", "self_knowledge_accuracy", "social_recognition", "external_devaluation"]),
    "self_knowledge_accuracy": ols(df, "self_knowledge_accuracy", ["social_recognition", "external_devaluation", "self_concept_positivity", "self_esteem"]),
    "note": "Synthetic demonstration only; not clinical assessment, identity evaluation, personality testing, workplace screening, legal evaluation, or individual prediction.",
}

corr_cols = BASE + DERIVED
context_summary.to_csv(OUT / "python_self_system_context_summary.csv", index=False)
agreement_summary.to_csv(OUT / "python_self_other_agreement_summary.csv", index=False)
profile_summary.to_csv(OUT / "python_self_system_profile_summary.csv", index=False)
pattern_summary.to_csv(OUT / "python_self_system_pattern_summary.csv", index=False)
df[corr_cols].corr().to_csv(OUT / "python_self_system_correlations.csv")
(OUT / "python_self_system_models.json").write_text(json.dumps(models, indent=2))
df.to_csv(OUT / "python_scored_self_concept_self_esteem_self_knowledge.csv", index=False)
print(f"Wrote outputs to {OUT}")
